[pytree] Add custom treespec fqn field

angelayi opened the pull request:
Custom classes that are serialized with pytree are serialized by default with `f”{class.__module__}.{class.__name__}”`. This is a dependency from our serialized program directly into the outer Python environment. If a user moves the class to a different directory, the serialized program will be unable to be loaded. So, we will require users to pass in an FQN if they want to serialize their custom treespec type.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112428



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng @avikchaudhuri @gmagogsfm @zhxchen17 @tugsbayasgalan @suo

Differential Revision: [D50886366](https://our.internmc.facebook.com/intern/diff/D50886366)

angelayi committed:
[pytree] Add custom treespec fqn field

[ghstack-poisoned]

suo commented on torch/utils/_pytree.py:
```diff
@@ -599,6 +607,8 @@ def _treespec_to_json(spec: TreeSpec) -> _TreeSpecSchema:
     if spec.type not in SUPPORTED_SERIALIZED_TYPES:
         raise NotImplementedError(
             f"Serializing {spec.type} in pytree is not registered."
+            "You can register using torch.utils._pytree._register_pytree_node "
```
This error message is correct, but nevertheless I can imagine this being a little confusing to users ("but I *did* register my pytree!!"), who may not parse all this business with fqns as being the crucial element.

To make this more instantly fixable by users, I suggest: if `type_fqn is None`  on L142 above, then instead of skipping adding to `SUPPORTED_SERIALIZED_TYPES`, we serialize a dummy `NoTypeFqnFound` type. Then, in the error message here, we can check for the presence of that type, and offer a much more direct error message, e.g. "please update your _register_pytree_node call with a `type_fqn` kwarg".

I think this takes the percentage of cases that will need direct support from us from 20% to like 1%, which is nice.

suo commented on torch/utils/_pytree.py:
```diff
@@ -97,6 +97,7 @@ def _register_pytree_node(
     to_str_fn: Optional[ToStrFunc] = None,
     maybe_from_str_fn: Optional[MaybeFromStrFunc] = None,
     *,
+    type_fqn: Optional[str] = None,
```
`type_fqn` seems a little opaque as name for this concept. A user may even be tempted to set this to the type's fqn, which would be wrong lol. 

Maybe: `serialized_type_name`?

suo commented on torch/utils/_pytree.py:
```diff
@@ -136,12 +139,12 @@ def _register_pytree_node(
+            typ, type_fqn, to_dumpable_context, from_dumpable_context
+        )
+        SUPPORTED_SERIALIZED_TYPES[typ] = serialize_node_def
+        SERIALIZED_TYPE_TO_PYTHON_TYPE[type_fqn] = typ
```
What should the behavior be on duplicate registrations? Right now it is "last one wins", but perhaps we should explicitly error?

angelayi replied:
I think in the near future they will disable duplicate registrations: https://github.com/pytorch/pytorch/pull/112111/files#r1376411179

zou3519 commented on test/test_pytree.py:
```diff
@@ -543,6 +543,7 @@ def __init__(self, x, y):
             DummyType,
             lambda dummy: ([dummy.x, dummy.y], None),
             lambda xs, _: DummyType(*xs),
+            type_fqn="test_pytree_custom_type_serialize.DummyType",
```
Why doesn't python pickle have this problem?

suo replied:
it does

angelayi committed:
Update on "[pytree] Add custom treespec fqn field"

Custom classes that are serialized with pytree are serialized by default with `f”{class.__module__}.{class.__name__}”`. This is a dependency from our serialized program directly into the outer Python environment. If a user moves the class to a different directory, the serialized program will be unable to be loaded. So, we will require users to pass in an FQN if they want to serialize their custom treespec type.




[ghstack-poisoned]

angelayi committed:
Update on "[pytree] Add custom treespec fqn field"

Custom classes that are serialized with pytree are serialized by default with `f”{class.__module__}.{class.__name__}”`. This is a dependency from our serialized program directly into the outer Python environment. If a user moves the class to a different directory, the serialized program will be unable to be loaded. So, we will require users to pass in an FQN if they want to serialize their custom treespec type.




cc avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan suo

[ghstack-poisoned]

angelayi commented:
@angelayi has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50886366).

zou3519 commented on torch/utils/_pytree.py:
```diff
@@ -137,6 +139,8 @@ def _register_pytree_node(
             flattened pytree.
         unflatten_fn: A callable that takes a flattened version of the pytree,
             additional context, and returns an unflattedn pytree.
+        serialized_type_name: A keyword argument used to specify the fully qualified
+            name used when serializing the tree spec.
```
We've added a lot of fields to register_pytree_node. I've noticed that users have attempted to provide all of the things we ask for here.

We should make it clear (by adding a line above "Args" in the docstring) that the only required fields are typ, flatten_fn, and unflatten_fn. The other fields are optional and only necessary if someone wants to json serialize their type

angelayi committed:
Update on "[pytree] Add custom treespec fqn field"

Custom classes that are serialized with pytree are serialized by default with `f”{class.__module__}.{class.__name__}”`. This is a dependency from our serialized program directly into the outer Python environment. If a user moves the class to a different directory, the serialized program will be unable to be loaded. So, we will require users to pass in an FQN if they want to serialize their custom treespec type.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan suo

Differential Revision: [D50886366](https://our.internmc.facebook.com/intern/diff/D50886366)

[ghstack-poisoned]

angelayi committed:
Update on "[pytree] Add custom treespec fqn field"

Custom classes that are serialized with pytree are serialized by default with `f”{class.__module__}.{class.__name__}”`. This is a dependency from our serialized program directly into the outer Python environment. If a user moves the class to a different directory, the serialized program will be unable to be loaded. So, we will require users to pass in an FQN if they want to serialize their custom treespec type.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng avikchaudhuri gmagogsfm zhxchen17 tugsbayasgalan suo

Differential Revision: [D50886366](https://our.internmc.facebook.com/intern/diff/D50886366)

[ghstack-poisoned]

huydhn commented:
If you need to revert this, the best way would be to ninja unland the internal diff.  That will be reflected here automatically.

angelayi commented:
@huydhn I have a forward fix here: https://github.com/pytorch/pytorch/pull/112748. Waiting on CI, but if you think CI is good I can try to force merge again
